src/box/box.cc

```diff
@@ -793,6 +798,80 @@ wal_stream_apply_dml_row(struct wal_stream *stream, struct xrow_header *row)
 	return -1;
 }
 
+struct rlist nodes_rows[VCLOCK_MAX];
+
+/*
+ * When restoring the log is read row-by-row. However, it is necessary
+ * to store the rows for correct further recovery. For example, with
+ * mixed transactions. The following function saves the newly arrived
+ * row to the rlist, which is stored in array is nodes_rows in cell is
+ * replica_id-1. As soon as a row arrives with the is_commit flag set,
+ * the corresponding transaction tries to apply. If an error occurs the
+ * transaction is rolled back. And we continue to work with the next
+ * transaction.
+ */
+static int
+wal_stream_apply_mixed_dml_row(struct wal_stream *stream,
+			       struct xrow_header *row)
+{
+	int rc = 0;
+	/* replica_id started from 1 */
+	uint32_t id = row->replica_id - 1;
+	/* The rlist can be created if the cell of array is empty */
+	if (nodes_rows[id].prev == nodes_rows[id].next &&
+	    nodes_rows[id].prev == NULL) {
+		rlist_create(&nodes_rows[id]);
+		assert(rlist_empty(&nodes_rows[id]));
+	}
+
+	/* When a new row arrives, it has the same address. Because the
+	 * row is on the stack. Therefore, it needs to allocate memory on
+	 * the heap. We do with row->body[0].iov_base analogically.
+	 */
+	struct xrow_header *alloc_row = (struct xrow_header *)malloc(sizeof(*alloc_row));
+	if (alloc_row == NULL)
+		tnt_raise(OutOfMemory, sizeof(*alloc_row), "malloc",
+			  "alloc_row");
+	memcpy(alloc_row, row, sizeof(*row));
+
+	char *alloc_body = (char *)malloc(row->body[0].iov_len);
+	if (alloc_body == NULL)
+		tnt_raise(OutOfMemory, sizeof(*alloc_body), "malloc",
+			  "alloc_body");
+	memcpy(alloc_body, row->body[0].iov_base, row->body[0].iov_len);
+	alloc_row->body[0].iov_base = alloc_body;
+
+	rlist_add_tail_entry(&nodes_rows[id], alloc_row, in_row_list);
+
+	if (!row->is_commit) {
+		return 0;
+	} else {
+		struct xrow_header *item, *next_row;
+		rlist_foreach_entry_safe(item, &nodes_rows[id],
+					 in_row_list, next_row) {
+			rlist_del_entry(item, in_row_list);
+			if (rc == 0) {
+				rc = wal_stream_apply_dml_row(stream, item);
+			}
+			/* deallocating the memory */
+			alloc_body = (char *)item->body[0].iov_base;
+			free(alloc_body);
+			alloc_body = NULL;
+			alloc_row = item;
+			free(alloc_row);
+			alloc_row = NULL;
+		}
+		/* clear the cell of array */
+		assert(rlist_empty(&nodes_rows[id]));
+		assert(nodes_rows[id].prev != NULL);
+		nodes_rows[id].prev = NULL;
+		assert(nodes_rows[id].next != NULL);
+		nodes_rows[id].next = NULL;
```

Comment:
There's no need to clear the list extensively. Having an empty rlist (which has `rlist_empty(...)` == true) is totally fine. It's ready to accept new members.

Reply:
I create a list if `nodes_rows[id].prev == NULL` and if `nodes_rows[id].next == NULL`. Therefore, checking that the list is empty is not enough. I don’t see the point of using  `RLIST_HEAD(nodes_rows[id]) ` macro, because it will complicate the logic for an array.